### ISSUE TITLE
LeaderElector stuck infinitely if configuration is missing or wrong

### ISSUE THREAD

**Issue by ST-DDT** [kubernetes-client/java#1011]
The leader elector gets stuck when it is unable to aquire the lease and spams the console with errors.

````txt
15:04:23.201 [main] DEBUG io.kubernetes.client.util.ClientBuilder - Could not find file specified in $KUBECONFIG
15:04:23.207 [main] DEBUG io.kubernetes.client.util.ClientBuilder - Could not find ~/.kube/config
15:04:34.220 [main] INFO  io.kubernetes.client.extended.leaderelection.LeaderElector - Start leader election with lock kube-system/leader-election
15:04:45.034 [main] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Attempting to acquire leader lease...
15:04:47.084 [leader-elector-scheduled-worker-0] ERROR io.kubernetes.client.extended.leaderelection.LeaderElector - Error processing tryAcquireOrRenew as null
15:04:49.123 [leader-elector-lease-worker-0] ERROR io.kubernetes.client.extended.leaderelection.LeaderElector - Error retrieving resource lock kube-system/leader-election as java.net.ConnectException: Failed to connect to localhost/0:0:0:0:0:0:0:1:8080
15:04:51.581 [leader-elector-scheduled-worker-0] ERROR io.kubernetes.client.extended.leaderelection.LeaderElector - Error processing tryAcquireOrRenew as null
15:04:53.617 [leader-elector-lease-worker-0] ERROR io.kubernetes.client.extended.leaderelection.LeaderElector - Error retrieving resource lock kube-system/leader-election as java.net.ConnectException: Failed to connect to localhost/0:0:0:0:0:0:0:1:8080
15:04:56.078 [leader-elector-scheduled-worker-0] ERROR io.kubernetes.client.extended.leaderelection.LeaderElector - Error processing tryAcquireOrRenew as null
````

It would be nice if there was some kind of error callback, that can be used to quit the inifinite loop and set the app health status to unhealthy/kill the application.

https://github.com/kubernetes-client/java/blob/9211be5f8bfc34328e277379bfbfa412dbe14898/examples/src/main/java/io/kubernetes/client/examples/LeaderElectionExample.java

* kubernetes `client-java-extended` v8.0.2
* Java 11 (64bit)
* Windows 10

**Comment by yue9944882** [comment]
i was working on adding error handler for those workers running in background but it's still in my queue, there's something similar newly introduced to the golang's client library https://github.com/kubernetes/client-go/blob/db5293afc7bf338a58c5b9992ed416b187642c2e/tools/cache/shared_informer.go#L182. i will send a refactor to pick the change to the java library..

/assign

**Comment by ST-DDT** [comment]
While you are already at it, please also consider adding the following:
- continuous mode (leader -> replica -> leader), without re-registering/re-running
- "is-replica"-notification (after startup)
- stepDown() functionality (before a scheduled shutdown or if the pods notices that it isn't healthy)
- pause()/continue()
- stop()/close() functionality, stop the execution loop without becoming leader (for one time tasks).
- clean() functionality (for one time tasks)

If you are busy/these issues have low priority, I'm willing to contribute (might take a few days though).
Not sure whether its preferable to keep the existing API with all these changes or to break it.

**Comment by yue9944882** [comment]
@ST-DDT thanks, here's my 2 cents on your suggestions:

> continuous mode (leader -> replica -> leader), without re-registering/re-running

am +1 on this. for clarification the so-called "continuous-mode" is not how the original leader-election works in the upstream golang library. but that's given the fact that golang application is more light-weight, the whole application can simply exit then restart if it loses the lock. however, restarting is usually more costy for java application..

> "is-replica"-notification (after startup)

can you elaborate on "is-replicas"?

> stepDown() functionality (before a scheduled shutdown or if the pods notices that it isn't healthy)

there's already a pre-stop callback to do the preppings before shutting down.. is that what you're looking for?

https://github.com/kubernetes-client/java/blob/33b9f0187aa92a48eb805555875d544589dfef35/examples/src/main/java/io/kubernetes/client/examples/LeaderElectionExample.java#L40-L42

> pause()/continue()
> stop()/close() functionality, stop the execution loop without becoming leader (for one time tasks).

if i understand correctly, all the proposed functions above enables the controller to be working "proactively" (i.e. users can interact w/ a running controller). however i presume the controller is designed to be reactive, working completely driven by the watch events from the kubernetes cluster. in order to suspend the controller from running, users are supposed to achieve it by scaling the pods/services.

> clean() functionality (for one time tasks)

by `one time tasks`, are you talking about listing all resources then processing them once? if that's the idea, you can do it by setting a non-zero [global resyncPeriod](https://github.com/kubernetes-client/java/blob/58ea383381c67a7494cdffc72457e327463f7d38/util/src/main/java/io/kubernetes/client/informer/SharedInformerFactory.java#L123) or [per-watch-handler resyncPeriod](https://github.com/kubernetes-client/java/blob/58ea383381c67a7494cdffc72457e327463f7d38/util/src/main/java/io/kubernetes/client/informer/SharedInformer.java#L23)

**Comment by yue9944882** [comment]
> If you are busy or these issues have low priority for you, I'm willing to contribute (might take a few days though).
Not sure whether its preferable to keep the existing (java-leader-election) API with all these changes or to break it.

i will be working on it if the ideas look good.. on the other hand, any contributions are always welcome!

**Comment by yue9944882** [comment]
@ST-DDT opened https://github.com/kubernetes-client/java/pull/1026 to support customized error handling, plz feel free to drop your comments, thanks

**Comment by ST-DDT** [comment]
> > "is-replica"-notification (after startup)
> 
> can you elaborate on "is-replicas"?

A notification like "sorry, the election was successful, but X became the leader".
Mainly intended for logging and exposing as info. Belongs to the category of explicit application state: NOT_PARTICIPATED (never run), LEADER, REPLICA, ERROR, FINISHED (for jobs).

Maybe change the two callbacks to a single one `(ElectionResult result, String leaderId) -> ...`:

- ElectionResult:
  - LEADER
  - REPLICA
  - ERROR

> > stepDown() functionality (before a scheduled shutdown or if the pods notices that it isn't healthy)
> 
> there's already a pre-stop callback to do the preppings before shutting down.. is that what you're looking for?
> 
> https://github.com/kubernetes-client/java/blob/33b9f0187aa92a48eb805555875d544589dfef35/examples/src/main/java/io/kubernetes/client/examples/LeaderElectionExample.java#L40-L42

The leader notices, that it is busy/no longer healthy/will be shutdown soon, but its lease is still valid for some time, so it explicitly steps down from being the leader (by deleting the lock?), allowing other instances to get the lease/becoming the leader earlier.

It works roughly like this:

1. Stop being leader
2. Wait `>1.0` intervals for other instances to become leader
3. Participate in the election again (unless frozen or stopped)

https://docs.mongodb.com/manual/reference/method/rs.stepDown/

> > pause()/continue()
> > stop()/close() functionality, stop the execution loop without becoming leader (for one time tasks).
> 
> if i understand correctly, all the proposed functions above enables the controller to be working "proactively" (i.e. users can interact w/ a running controller). however i presume the controller is designed to be reactive, working completely driven by the watch events from the kubernetes cluster. in order to suspend the controller from running, users are supposed to achieve it by scaling the pods/services.

No, its still related for the leader election.

`pause()`/`continue()` are related to election status refreshs. (Do not try to refresh your leader lease or try to become leader)
Usually if the service is temporarily not ready/busy/not healthy, the instance should stop trying to become leader.
(`pause()` might trigger explicit `stepDown()`)

https://docs.mongodb.com/manual/reference/method/rs.freeze/
(But not time based, but on/off based)

`close()`(`stop()`): Closes the scheduledWorkers, leaseWorkers, hookWorkers resources to avoid `Thread` leaks.
(Implies `stepDown()`)

> > clean() functionality (for one time tasks)
> 
> by `one time tasks`, are you talking about listing all resources then processing them once? if that's the idea, you can do it by setting a non-zero [global resyncPeriod](https://github.com/kubernetes-client/java/blob/58ea383381c67a7494cdffc72457e327463f7d38/util/src/main/java/io/kubernetes/client/informer/SharedInformerFactory.java#L123) or [per-watch-handler resyncPeriod](https://github.com/kubernetes-client/java/blob/58ea383381c67a7494cdffc72457e327463f7d38/util/src/main/java/io/kubernetes/client/informer/SharedInformer.java#L23)

`clean()`/ `purge()` removes the resource created by the lock (Endpoint/ConfigMap).
I use this in tests to cleanup after/between them.
And I also consider using it in combination with the `is-replica` notification for "long running" generation tasks.
The report generation takes 15-30 Seconds, so the container that received the requests creates such a lock. Any other instance that will try to generate it (due to user impatience or just coincidental parallelism) will notice "oh the lock exists already" -> respond with WIP. After the generation is done the lock needs to be removed eventually, to avoid flooding the kubernetes with dead elements. I can do so via the resource API, but since I created it via the elector, it should be possible to delete it there as well.

> > If you are busy or these issues have low priority for you, I'm willing to contribute (might take a few days though).
> > Not sure whether its preferable to keep the existing (java-leader-election) API with all these changes or to break it.
> 
> i will be working on it if the ideas look good.. on the other hand, any contributions are always welcome!

Since my suggestions require a massive rework of the class, working in parallel on it isn't possible/worth it.
However, I might create a draft/PR to better express my suggestion.

**Comment by yue9944882** [comment]
> A notification like "sorry, the election was successful, but X became the leader".
Mainly intended for logging and exposing as info. Belongs to the category of explicit application state: NOT_PARTICIPATED (never run), LEADER, REPLICA, ERROR, FINISHED (for jobs).

@ST-DDT can you take a look at https://github.com/kubernetes-client/java/pull/1031? the pull is picking something from the golang library upstream and it's probably resolving your case.

as for the other extended capabilities like `stepDown()`/`clean()`, am not sure if it's helpful for enhancing operator but it's good to be a opt-in extension.. plz feel free to send PRs

**Comment by ST-DDT** [comment]
The PR looks good to me. However, I noticed some strange behavior (not related to the PR it just increases the visibility).

It can easily be reproduced by using the LeaderElectionExample.

App-Run1-Logs:
````txt
09:58:47.340 [main] INFO  io.kubernetes.client.extended.leaderelection.LeaderElector - Start leader election with lock kube-system/leader-election
09:58:47.345 [main] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Attempting to acquire leader lease...
09:58:47.580 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Lock is held by a78c7c58-34fa-4cb8-8153-81a4bdb04d08 and has not yet expired: 2020-07-02T09:56:06.847+0200
09:58:47.585 [leader-elector-scheduled-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - The tryAcquireOrRenew result is false

Leader: a78c7c58-34fa-4cb8-8153-81a4bdb04d08

09:58:50.916 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Lock is held by a78c7c58-34fa-4cb8-8153-81a4bdb04d08 and has not yet expired: 2020-07-02T09:56:06.847+0200
09:58:50.917 [leader-elector-scheduled-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - The tryAcquireOrRenew result is false
09:58:54.242 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Lock is held by a78c7c58-34fa-4cb8-8153-81a4bdb04d08 and has not yet expired: 2020-07-02T09:56:06.847+0200
09:58:54.242 [leader-elector-scheduled-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - The tryAcquireOrRenew result is false
09:58:57.568 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Lock is held by a78c7c58-34fa-4cb8-8153-81a4bdb04d08 and has not yet expired: 2020-07-02T09:56:06.847+0200
09:58:57.568 [leader-elector-scheduled-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - The tryAcquireOrRenew result is false
09:59:00.894 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Update lock acquire time to keep lease
09:59:00.919 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - TryAcquireOrRenew return success
09:59:00.919 [leader-elector-scheduled-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - The tryAcquireOrRenew result is true

Leader: 0260b94e-d6d0-4c8f-9481-a63200f79016

09:59:01.350 [main] INFO  io.kubernetes.client.extended.leaderelection.LeaderElector - Successfully acquired lease, became leader
09:59:01.350 [main] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Attempting to renew leader lease...

Do something when getting leadership.

09:59:01.356 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Update lock acquire time to keep lease
09:59:01.362 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - TryAcquireOrRenew return success
09:59:01.362 [main] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Successfully renewed lease
09:59:03.368 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Update lock acquire time to keep lease
09:59:03.377 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - TryAcquireOrRenew return success
09:59:03.378 [main] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Successfully renewed lease
09:59:05.387 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Update lock acquire time to keep lease
09:59:05.395 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - TryAcquireOrRenew return success
09:59:05.396 [main] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Successfully renewed lease
09:59:07.405 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Update lock acquire time to keep lease
09:59:07.416 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - TryAcquireOrRenew return success
09:59:07.417 [main] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Successfully renewed lease
09:59:09.426 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Update lock acquire time to keep lease
09:59:09.436 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - TryAcquireOrRenew return success
09:59:09.437 [main] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Successfully renewed lease
````
App-Run2-Logs (started 5 Minutes later):
````txt
10:05:37.756 [main] INFO  io.kubernetes.client.extended.leaderelection.LeaderElector - Start leader election with lock kube-system/leader-election
10:05:37.761 [main] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Attempting to acquire leader lease...
10:05:38.002 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Lock is held by 0260b94e-d6d0-4c8f-9481-a63200f79016 and has not yet expired: 2020-07-02T09:59:09.418+0200
10:05:38.007 [leader-elector-scheduled-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - The tryAcquireOrRenew result is false

Leader: 0260b94e-d6d0-4c8f-9481-a63200f79016

10:05:41.447 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Lock is held by 0260b94e-d6d0-4c8f-9481-a63200f79016 and has not yet expired: 2020-07-02T09:59:09.418+0200
10:05:41.447 [leader-elector-scheduled-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - The tryAcquireOrRenew result is false
10:05:44.872 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Lock is held by 0260b94e-d6d0-4c8f-9481-a63200f79016 and has not yet expired: 2020-07-02T09:59:09.418+0200
10:05:44.873 [leader-elector-scheduled-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - The tryAcquireOrRenew result is false
10:05:48.295 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Update lock acquire time to keep lease
10:05:48.307 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - TryAcquireOrRenew return success
10:05:48.307 [leader-elector-scheduled-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - The tryAcquireOrRenew result is true

Leader: 1ea061b2-9a5c-4486-9d0e-13f1ab23c54e

10:05:49.766 [main] INFO  io.kubernetes.client.extended.leaderelection.LeaderElector - Successfully acquired lease, became leader

Do something when getting leadership.

10:05:49.766 [main] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Attempting to renew leader lease...
10:05:49.778 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Update lock acquire time to keep lease
10:05:49.786 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - TryAcquireOrRenew return success
10:05:49.787 [main] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Successfully renewed lease
10:05:51.796 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Update lock acquire time to keep lease
10:05:51.805 [leader-elector-lease-worker-0] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - TryAcquireOrRenew return success
10:05:51.806 [main] DEBUG io.kubernetes.client.extended.leaderelection.LeaderElector - Successfully renewed lease
````

The lease clearly should have been expired by then, but the newly started application still reports the old instance as leader for an entire cycle.

**Comment by newzubakhin** [comment]
/reopen

Hi. Is there a way to handle `ApiException` during leader election? Now it causes endless loop with exceptions.
